### ISSUE TITLE
Inconsistencies in enum unit vs empty tuple vs empty struct

### ISSUE THREAD

**Issue by vojtechkral** [rust-lang/rust#44653]
Hi,
  I'm running into couple of rough / weird corner cases with empty enum tuple and struct variants.
Code to demonstrate:

```rust

enum Example {
    Foo,
    Bar(),
    Baz{},
}

fn main() {
    // Foo:
    // let x = Example::Foo();
    // ^- error[E0618]: expected function, found `Example::Foo`
    let x = Example::Foo{};
    // ^- compiles

    // Bar:
    let x = Example::Bar;
    // ^- compiles
    let x = Example::Bar{};
    // ^- compiles

    // Baz:
    // let x = Example::Baz;
    // ^- error[E0423]: expected value, found struct variant `Example::Baz`
    // let x = Example::Baz();
    // ^- error[E0423]: expected function, found struct variant `Example::Baz`
}
```

As you can see, the behaviour here is not consistent.
For example, you can initialize an empty tuple variant with a syntax of unit variant, but not vice versa.
Also, the error messages are not super helpful.

The reason I ran into this is metaprogramming with both regular and procedural macros.
In macros, when generating code that initializes enums, one can't use a `Foo()` syntax to initialize `Foo` and needs to special-case to omit the parens.

Would it make sense to perhaps make all of `Foo`, `Foo()` and `Foo{}` mean the same, ie. unit?

**Comment by petrochenkov** [comment]
@vojtechkral 
The rules are described in https://github.com/rust-lang/rfcs/blob/master/text/1506-adt-kinds.md
TLDR: `Foo = Foo {} + const Foo`, `Bar() = Bar {} + fn Bar`.

So, `Foo {}` syntax can be used for everything without special cases.
`Example::Foo()` is a usual function call, it doesn't compile because `Example::Foo` is not a function.
`let x = Example::Bar;` compiles, but is actually a function item, not `Example `, you can call it `x()`.
`Example::Baz` and `Example::Baz()` don't compile because `Baz {}` doesn't create neither constant nor function.

**Comment by vojtechkral** [comment]
@petrochenkov  Ah, ok, thanks for explanation. I guess as long as the semantics are well-defined, it's okay, although it seems a bti confusing to me...

I noticed I can also use the `{}` syntax to initialize tuple variants even when they have actual items, ie.

```rust
let x = Example::Foo{ 0: somvalue , 1: someothervalue };
```

Is this well-defined and ok to use this syntax? This way my macro code could get simplified...

**Comment by steveklabnik** [comment]
Given that this isn't a bug, I'm going to give this a close.

**Comment by vojtechkral** [comment]
@steveklabnik Yeah, sure, I forgot to close, sorry...
